lib/ansible/module_utils/redfish_utils.py

```diff
@@ -1566,4 +1575,107 @@ def get_system_inventory(self, systems_uri):
+            if r.get('ret'):
+                d = r.get('data')
+            else:
+                return
+
+        if 'Members' in d:  # collections case
+            for m in d.get('Members'):
+                u = m.get('@odata.id')
+                s = self.get_request(self.root_uri + u)
+                if s.get('ret'):
+                    p = s.get('data')
+                    if p:
+                        e = {subsystem.lower() + '_uri': u,
+                             status: p.get(status,
+                                           "Status not available")}
+                        health[subsystem].append(e)
+        else:  # non-collections case
+            e = {subsystem.lower() + '_uri': uri,
+                 status: d.get(status,
+                               "Status not available")}
+            health[subsystem].append(e)
+
+    def get_health_subsystem(self, subsystem, data, health):
+        if subsystem in data:
+            health[subsystem] = []
+            sub = data.get(subsystem)
+            if isinstance(sub, list):
+                for r in sub:
+                    if '@odata.id' in r:
+                        uri = r.get('@odata.id')
+                        expanded = None
+                        if '#' in uri and len(r) > 1:
+                            expanded = r
+                        self.get_health_resource(subsystem, uri, health, expanded)
+            elif isinstance(sub, dict):
+                if '@odata.id' in sub:
+                    uri = sub.get('@odata.id')
+                    self.get_health_resource(subsystem, uri, health, None)
+
+    def get_health_report(self, category, uri, subsystems):
+        result = {}
+        health = {}
+        status = 'Status'
+
+        # Get health status of top level resource
+        response = self.get_request(self.root_uri + uri)
+        if response['ret'] is False:
+            return response
+        result['ret'] = True
+        data = response['data']
+        health[category] = {status: data.get(status, "Status not available")}
+
+        # Get health status of subsystems
+        for sub in subsystems:
+            d = None
+            if sub.startswith('Links.'):  # ex: Links.PCIeDevices
+                sub = sub[len('Links.'):]
+                d = data.get('Links', {})
+            elif '.' in sub:  # ex: Thermal.Fans
+                p, sub = sub.split('.')
+                u = data.get(p, {}).get('@odata.id')
+                if u:
+                    r = self.get_request(self.root_uri + u)
+                    if r['ret']:
+                        d = r['data']
+                if not d:
+                    continue
+            else:  # ex: Memory
+                d = data
+            self.get_health_subsystem(sub, d, health)
+
+        result["entries"] = health
+        return result
+
+    def get_system_health_report(self, systems_uri):
+        subsystems = ['Processors', 'Memory', 'Storage', 'EthernetInterfaces']
```

Comment:
Might want to add SimpleStorage and NetworkInterfaces

Reply:
Added in new commit.